In [ ]:
!pip install segmentation_models_pytorch torchviz torchinfo torchsummary

https://medium.com/@AriaLeeNotAriel/numbynum-segformer-simple-and-efficient-design-for-semantic-segmentation-with-transformers-29c1740e04ce

https://towardsdatascience.com/implementing-segformer-in-pytorch-8f4705e2ed0e

https://github.com/NVlabs/SegFormer/blob/master/mmseg/models/backbones/mix_transformer.py

https://arxiv.org/pdf/2105.15203

In [ ]:
import segmentation_models_pytorch as smp

common_args = {"encoder_name": "mit_b0",
               "encoder_depth": 5,
               "encoder_weights": "imagenet",
               "decoder_dropout": 0.3,
               "in_channels": 3,
               "classes": 19,
               "activation": None}
model = smp.FPN(**common_args)

Downloading: "https://github.com/qubvel/segmentation_models.pytorch/releases/download/v0.0.2/mit_b0.pth" to /root/.cache/torch/hub/checkpoints/mit_b0.pth
100%|██████████| 13.7M/13.7M [00:00<00:00, 116MB/s] 


In [ ]:
model

FPN(
  (encoder): MixVisionTransformerEncoder(
    (patch_embed1): OverlapPatchEmbed(
      (proj): Conv2d(3, 32, kernel_size=(7, 7), stride=(4, 4), padding=(3, 3))
      (norm): LayerNorm((32,), eps=1e-05, elementwise_affine=True)
    )
    (patch_embed2): OverlapPatchEmbed(
      (proj): Conv2d(32, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1))
      (norm): LayerNorm((64,), eps=1e-05, elementwise_affine=True)
    )
    (patch_embed3): OverlapPatchEmbed(
      (proj): Conv2d(64, 160, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1))
      (norm): LayerNorm((160,), eps=1e-05, elementwise_affine=True)
    )
    (patch_embed4): OverlapPatchEmbed(
      (proj): Conv2d(160, 256, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1))
      (norm): LayerNorm((256,), eps=1e-05, elementwise_affine=True)
    )
    (block1): ModuleList(
      (0): Block(
        (norm1): LayerNorm((32,), eps=1e-06, elementwise_affine=True)
        (attn): Attention(
          (q): Linear(in_features=32, ou

In [ ]:
from torchviz import make_dot
import torch

# Create a dummy input
dummy_input = torch.randn(1, 3, 1024, 2048)  # Batch size 1, 3 channels, 256x256 resolution

# Generate computation graph
output = model(dummy_input)
graph = make_dot(output, params=dict(model.named_parameters()))

# Save the computation graph as a file
graph.render("FPN_mitb3_model_computation_graph", format="pdf")

'FPN_mitb3_model_computation_graph.pdf'

In [ ]:
from torchinfo import summary

# Generate and print a model summary
summary(model, input_size=(1, 3, 1024, 2048))  # Batch size 1, input 3 channels, 256x256 resolution

Layer (type:depth-idx)                                  Output Shape              Param #
FPN                                                     [1, 19, 1024, 2048]       --
├─MixVisionTransformerEncoder: 1-1                      [1, 3, 1024, 2048]        --
│    └─OverlapPatchEmbed: 2-1                           [1, 131072, 32]           --
│    │    └─Conv2d: 3-1                                 [1, 32, 256, 512]         4,736
│    │    └─LayerNorm: 3-2                              [1, 131072, 32]           64
│    └─ModuleList: 2-2                                  --                        --
│    │    └─Block: 3-3                                  [1, 131072, 32]           79,616
│    │    └─Block: 3-4                                  [1, 131072, 32]           79,616
│    └─LayerNorm: 2-3                                   [1, 131072, 32]           64
│    └─OverlapPatchEmbed: 2-4                           [1, 32768, 64]            --
│    │    └─Conv2d: 3-5                          

In [2]:
!pip install datasets transformers

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 480.6/480.6 kB 27.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 9.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 16.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 10.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 16.8 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2024.10.0
    Uninstalling fsspec-2024.10.0:
      Successfully uninstalled fsspec-2024.10.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2024.10.0 requires fsspec==2024.10.0, but you have fsspec 2024.9.0 which is incompatible.


In [3]:
!pip install kaleido

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.9/79.9 MB 29.2 MB/s eta 0:00:00


In [4]:
from huggingface_hub import login
from google.colab import userdata

login(token=userdata.get('HF_TOKEN'))

In [ ]:
import datasets, transformers
from datasets import load_dataset

cityscapes_semantic = load_dataset("BhavanaMalla/cityscapes-semantic")

README.md:   0%|          | 0.00/654 [00:00<?, ?B/s]

train-00000-of-00014.parquet:   0%|          | 0.00/484M [00:00<?, ?B/s]

train-00001-of-00014.parquet:   0%|          | 0.00/518M [00:00<?, ?B/s]

train-00002-of-00014.parquet:   0%|          | 0.00/500M [00:00<?, ?B/s]

train-00003-of-00014.parquet:   0%|          | 0.00/491M [00:00<?, ?B/s]

train-00004-of-00014.parquet:   0%|          | 0.00/472M [00:00<?, ?B/s]

train-00005-of-00014.parquet:   0%|          | 0.00/509M [00:00<?, ?B/s]

train-00006-of-00014.parquet:   0%|          | 0.00/526M [00:00<?, ?B/s]

train-00007-of-00014.parquet:   0%|          | 0.00/502M [00:00<?, ?B/s]

train-00008-of-00014.parquet:   0%|          | 0.00/491M [00:00<?, ?B/s]

train-00009-of-00014.parquet:   0%|          | 0.00/512M [00:00<?, ?B/s]

train-00010-of-00014.parquet:   0%|          | 0.00/510M [00:00<?, ?B/s]

train-00011-of-00014.parquet:   0%|          | 0.00/513M [00:00<?, ?B/s]

train-00012-of-00014.parquet:   0%|          | 0.00/515M [00:00<?, ?B/s]

train-00013-of-00014.parquet:   0%|          | 0.00/510M [00:00<?, ?B/s]

validation-00000-of-00003.parquet:   0%|          | 0.00/416M [00:00<?, ?B/s]

validation-00001-of-00003.parquet:   0%|          | 0.00/398M [00:00<?, ?B/s]

validation-00002-of-00003.parquet:   0%|          | 0.00/386M [00:00<?, ?B/s]

test-00000-of-00007.parquet:   0%|          | 0.00/466M [00:00<?, ?B/s]

test-00001-of-00007.parquet:   0%|          | 0.00/492M [00:00<?, ?B/s]

test-00002-of-00007.parquet:   0%|          | 0.00/498M [00:00<?, ?B/s]

test-00003-of-00007.parquet:   0%|          | 0.00/501M [00:00<?, ?B/s]

test-00004-of-00007.parquet:   0%|          | 0.00/496M [00:00<?, ?B/s]

test-00005-of-00007.parquet:   0%|          | 0.00/502M [00:00<?, ?B/s]

test-00006-of-00007.parquet:   0%|          | 0.00/509M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/2975 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/500 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1525 [00:00<?, ? examples/s]

In [ ]:
cityscapes_semantic

DatasetDict({
    train: Dataset({
        features: ['image', 'annotation_labelids', 'annotation_labeltrainids', 'image_name'],
        num_rows: 2975
    })
    validation: Dataset({
        features: ['image', 'annotation_labelids', 'annotation_labeltrainids', 'image_name'],
        num_rows: 500
    })
    test: Dataset({
        features: ['image', 'annotation_labelids', 'annotation_labeltrainids', 'image_name'],
        num_rows: 1525
    })
})

In [ ]:
import plotly.graph_objects as go
from datasets import load_dataset
import numpy as np

# Load the Cityscapes dataset
cityscapes_semantic = load_dataset("BhavanaMalla/cityscapes-semantic")

# Define the color map and class names
class_names = [
    "Road", "Sidewalk", "Building", "Wall", "Fence", "Pole", "Traffic light", "Traffic sign",
    "Vegetation", "Terrain", "Sky", "Person", "Rider", "Car", "Truck", "Bus", "Train", "Motorcycle", "Bicycle"
]
palette = np.array([
    (128, 64, 128),  # Road
    (244, 35, 232),  # Sidewalk
    (70, 70, 70),    # Building
    (102, 102, 156), # Wall
    (190, 153, 153), # Fence
    (153, 153, 153), # Pole
    (250, 170, 30),  # Traffic light
    (220, 220, 0),   # Traffic sign
    (107, 142, 35),  # Vegetation
    (152, 251, 152), # Terrain
    (70, 130, 180),  # Sky
    (220, 20, 60),   # Person
    (255, 0, 0),     # Rider
    (0, 0, 142),     # Car
    (0, 0, 70),      # Truck
    (0, 60, 100),    # Bus
    (0, 80, 100),    # Train
    (0, 0, 230),     # Motorcycle
    (119, 11, 32)    # Bicycle
]) / 255.0

# Extract the pixel distribution
pixel_counts = {}

for sample in cityscapes_semantic["train"]:
    mask = sample["annotation_labeltrainids"]  # Access the segmentation mask
    unique, counts = np.unique(mask, return_counts=True)
    for cls, count in zip(unique, counts):
        if cls in pixel_counts:
            pixel_counts[cls] += count
        else:
            pixel_counts[cls] = count

# Prepare data for plotting
classes = [class_names[cls] for cls in pixel_counts.keys() if cls < len(class_names)]
pixels = [pixel_counts[cls] for cls in pixel_counts.keys() if cls < len(class_names)]

# Convert pixel counts to powers of 10
pixels_log = [np.log10(p) for p in pixels]


# Map colors to classes based on the pixel count keys
colors = []
for key in pixel_counts.keys():
    if key == 255:
        # Assign black color for key 255
        colors.append(f'rgb(0, 0, 0)')
    elif key < len(class_names):
        # Map the key to the corresponding color in the palette
        r, g, b = palette[key]
        colors.append(f'rgb({r*255}, {g*255}, {b*255})')





In [ ]:
pixel_counts

{0: 2036048947,
 1: 336030285,
 2: 1259773717,
 5: 67771817,
 7: 30522367,
 8: 878732742,
 9: 63964778,
 10: 221459205,
 11: 67202385,
 12: 7444903,
 13: 386502898,
 18: 22849764,
 255: 718433808,
 3: 36211195,
 4: 48487347,
 6: 11510397,
 17: 5445909,
 15: 12995799,
 14: 14775005,
 16: 12863932}

In [ ]:
classes

['Road',
 'Sidewalk',
 'Building',
 'Pole',
 'Traffic sign',
 'Vegetation',
 'Terrain',
 'Sky',
 'Person',
 'Rider',
 'Car',
 'Bicycle',
 'Wall',
 'Fence',
 'Traffic light',
 'Motorcycle',
 'Bus',
 'Truck',
 'Train']

In [ ]:
pixels

In [ ]:
pixels_log

In [ ]:
colors

In [ ]:
# Create the plot
fig = go.Figure(
    data=[
        go.Bar(
            x=classes,
            y=pixels_log,
            #text=[f'{p:.2e}' for p in pixels],
            text=[f'{p / 1e6:.2f}M' for p in pixels],
            textposition='auto',
            marker_color=colors
        )
    ]
)

# Update layout
fig.update_layout(
    title="Class Distribution in Cityscapes Dataset",
    xaxis_title="Classes",
    yaxis_title="Log10 of Pixel Counts",
    xaxis=dict(tickmode='array', tickvals=list(range(len(classes))), ticktext=classes),
    template="plotly_white"
)

fig.show()
# fig.write_image("class_distribution.png")

In [ ]:
sorted_indices = np.argsort(pixels)[::-1]
sorted_indices

array([ 0,  2,  5, 10,  1,  7,  3,  8,  6, 13, 12,  4, 11, 17, 16, 18, 14,
        9, 15])

In [ ]:
# Reorder pixels and classes according to the sorted indices
sorted_pixels = np.array(pixels)[sorted_indices]
sorted_classes = np.array(classes)[sorted_indices]
sorted_pixels, sorted_classes

(array([2036048947, 1259773717,  878732742,  386502898,  336030285,
         221459205,   67771817,   67202385,   63964778,   48487347,
          36211195,   30522367,   22849764,   14775005,   12995799,
          12863932,   11510397,    7444903,    5445909]),
 array(['Road', 'Building', 'Vegetation', 'Car', 'Sidewalk', 'Sky', 'Pole',
        'Person', 'Terrain', 'Fence', 'Wall', 'Traffic sign', 'Bicycle',
        'Truck', 'Bus', 'Train', 'Traffic light', 'Rider', 'Motorcycle'],
       dtype='<U13'))

In [ ]:
sorted_indices = np.argsort(pixels)[::-1]
sorted_pixels = np.array(pixels)[sorted_indices]
sorted_classes = np.array(classes)[sorted_indices]
sorted_colors = np.array(colors)[sorted_indices]

# Create the plot
fig = go.Figure(
    data=[go.Bar(
        x=sorted_classes,
        y=sorted_pixels,
        text=[f'{p / 1e6:.2f}M' for p in sorted_pixels],
        textposition='auto',
        marker_color=sorted_colors
    )]
)

# Update layout for logarithmic y-axis
fig.update_layout(
    # title="Class Distribution in Cityscapes Dataset",
    xaxis_title="Classes",
    yaxis_title="Pixel Counts",
    xaxis=dict(tickmode='array', tickvals=list(range(len(sorted_classes))), ticktext=sorted_classes),
    yaxis=dict(
        type='log',  # Use logarithmic scale for better visualization of small counts
        tickvals=[1, 10, 100, 1000, 10000, 100000, 1000000],  # Customize tick values as needed
        ticktext=[r'$10^0$', r'$10^1$', r'$10^2$', r'$10^3$', r'$10^4$', r'$10^5$', r'$10^6$']  # Powers of 10
    ),
    # template="plotly_white"
)

fig.show()
fig.write_image("class_distribution.pdf")

In [ ]:
import matplotlib.pyplot as plt
from datasets import load_dataset
import numpy as np

# Load the Cityscapes dataset
cityscapes_semantic = load_dataset("BhavanaMalla/cityscapes-semantic")

# Define the color map and class names
class_names = [
    "Road", "Sidewalk", "Building", "Wall", "Fence", "Pole", "Traffic light", "Traffic sign",
    "Vegetation", "Terrain", "Sky", "Person", "Rider", "Car", "Truck", "Bus", "Train", "Motorcycle", "Bicycle"
]
palette = np.array([
    (128, 64, 128),  # Road
    (244, 35, 232),  # Sidewalk
    (70, 70, 70),    # Building
    (102, 102, 156), # Wall
    (190, 153, 153), # Fence
    (153, 153, 153), # Pole
    (250, 170, 30),  # Traffic light
    (220, 220, 0),   # Traffic sign
    (107, 142, 35),  # Vegetation
    (152, 251, 152), # Terrain
    (70, 130, 180),  # Sky
    (220, 20, 60),   # Person
    (255, 0, 0),     # Rider
    (0, 0, 142),     # Car
    (0, 0, 70),      # Truck
    (0, 60, 100),    # Bus
    (0, 80, 100),    # Train
    (0, 0, 230),     # Motorcycle
    (119, 11, 32)    # Bicycle
]) / 255.0
ignored_color = (0, 0, 0)  # Gray for ignored regions

# Select one sample from the training dataset
sample = cityscapes_semantic["train"][96]
image = np.array(sample["image"])  # The input image
mask = np.array(sample["annotation_labeltrainids"])  # The ground truth mask

# Handle ignored regions in the mask
colored_mask = np.zeros((*mask.shape, 3))  # Initialize RGB mask
for cls_id in range(len(palette)):
    colored_mask[mask == cls_id] = palette[cls_id]
# Assign ignored color for regions with label 255
colored_mask[mask == 255] = ignored_color

# Plot and save the image
plt.figure(figsize=(10, 8))
plt.imshow(image)
plt.axis("off")
# plt.title("Cityscapes Image")
plt.savefig("cityscapes_4.pdf", bbox_inches="tight", pad_inches=0)
plt.close()

# Plot and save the mask
plt.figure(figsize=(10, 8))
plt.imshow(colored_mask)
plt.axis("off")
# plt.title("Ground Truth Mask")
plt.savefig("gt_mask_4.pdf", bbox_inches="tight", pad_inches=0)
plt.close()

print("PDFs saved: cityscapes_image.pdf and ground_truth_mask.pdf")


PDFs saved: cityscapes_image.pdf and ground_truth_mask.pdf


In [5]:
import datasets, transformers
from datasets import load_dataset

fishyscapes_semantic = load_dataset("BhavanaMalla/fishyscapes-semantic")

README.md:   0%|          | 0.00/347 [00:00<?, ?B/s]

data-00000-of-00001.parquet:   0%|          | 0.00/75.4M [00:00<?, ?B/s]

Generating data split:   0%|          | 0/30 [00:00<?, ? examples/s]

In [6]:
fishyscapes_semantic

DatasetDict({
    data: Dataset({
        features: ['image', 'mask', 'image_name'],
        num_rows: 30
    })
})

In [19]:
import os
from datasets import load_dataset

# Load the Fishyscapes dataset
fishyscapes_semantic = load_dataset("BhavanaMalla/fishyscapes-semantic")

# Directory to save images
output_dir = "downloaded_images"
os.makedirs(output_dir, exist_ok=True)

# Loop through the dataset and save PNG images
for sample in fishyscapes_semantic["data"]:
    image = sample["image"]  # Already a PNG
    image_name = sample["image_name"]

    # Save the image using the PIL Image save method
    image.save(os.path.join(output_dir, f"{image_name}.png"))

print(f"Images saved in {output_dir}")


Images saved in downloaded_images


In [ ]:
import matplotlib.pyplot as plt
from datasets import load_dataset
import numpy as np

# Load the Fishyscapes dataset
fishyscapes_semantic = load_dataset("BhavanaMalla/fishyscapes-semantic")

# Define the color map and class names
class_names = ["id", "ood", "background"]
palette = np.array([
    (70, 130, 180),  # ID - blues
    (220, 20, 60),   # OOD - red
    (0, 0, 0),       # Background - black
]) / 255.0

ignored_color = (0, 0, 0)
# Select one sample from the training dataset
sample = fishyscapes_semantic["data"][11] # 4, 11
image = np.array(sample["image"])  # The input image
mask = np.array(sample["mask"])  # The ground truth mask

# Handle ignored regions in the mask
colored_mask = np.zeros((*mask.shape, 3))  # Initialize RGB mask
for cls_id in range(len(palette)):
    colored_mask[mask == cls_id] = palette[cls_id]
# Assign ignored color for regions with label 255
colored_mask[mask == 255] = ignored_color

# Plot and save the image
plt.figure(figsize=(10, 8))
plt.imshow(image)
plt.axis("off")
plt.savefig("fishyscapes_2.pdf", bbox_inches="tight", pad_inches=0)
plt.close()

# Plot and save the mask
plt.figure(figsize=(10, 8))
plt.imshow(colored_mask)
plt.axis("off")
# plt.title("Ground Truth Mask")
plt.savefig("fishyscapes_gt_mask_2.pdf", bbox_inches="tight", pad_inches=0)
plt.close()

print("PDFs saved: fishyscapes_image.pdf and fishyscapesground_truth_mask.pdf")


PDFs saved: fishyscapes_image.pdf and fishyscapesground_truth_mask.pdf


In [ ]:
import matplotlib.pyplot as plt
from datasets import load_dataset
import numpy as np

# Load the Fishyscapes dataset
fishyscapes_semantic = load_dataset("BhavanaMalla/fishyscapes-semantic")

# Define the color map and class names
class_names = ["id", "ood", "background"]
palette = np.array([
    (5, 136, 245),  # ID - blues # (70, 130, 180)
    (220, 20, 60),   # OOD - red
    (0, 0, 0),       # Background - black
]) / 255.0

# Define ignored color
ignored_color = (0, 0, 0)

# Select one sample from the training dataset
sample = fishyscapes_semantic["data"][11]  # 4, 11 Change index as needed
image = np.array(sample["image"])  # The input image
mask = np.array(sample["mask"])  # The ground truth mask

# Create the transparent blue overlay
alpha = 0.5  # Transparency level (0: fully transparent, 1: fully opaque)

# Initialize the overlay image
overlay = image.copy() / 255.0  # Normalize the RGB image

id_mask = (mask == 0)  # Where the GT mask has "ID" class
# Apply the transparent blue overlay
overlay[id_mask] = alpha * palette[0] + (1 - alpha) * overlay[id_mask]

ood_mask = (mask == 1)  # Where the GT mask has "OOD" class
overlay[ood_mask] = alpha * palette[1] + (1 - alpha) * overlay[ood_mask]

bg_mask = (mask == 255)  # Where the GT mask has "Background" class
overlay[bg_mask] = palette[2]

# Plot and save the image
plt.figure(figsize=(10, 8))
plt.imshow(image)
plt.axis("off")
plt.savefig("fishyscapes_2.pdf", bbox_inches="tight", pad_inches=0)
plt.close()

# Plot and save the overlayed image
plt.figure(figsize=(10, 8))
plt.imshow(overlay)
plt.axis("off")
plt.savefig("fishyscapes_gt_mask_2.pdf", bbox_inches="tight", pad_inches=0)
plt.close()

print("fishyscapes image and gt_mask saved")


fishyscapes image and gt_mask saved


In [ ]:
import datasets, transformers
from datasets import load_dataset

lostandfound_semantic = load_dataset("BhavanaMalla/lostandfound-sub")

README.md:   0%|          | 0.00/351 [00:00<?, ?B/s]

data-00000-of-00001.parquet:   0%|          | 0.00/274M [00:00<?, ?B/s]

Generating data split:   0%|          | 0/100 [00:00<?, ? examples/s]

In [ ]:
lostandfound_semantic

DatasetDict({
    data: Dataset({
        features: ['image', 'mask', 'image_name'],
        num_rows: 100
    })
})

In [ ]:
import matplotlib.pyplot as plt
from datasets import load_dataset
import numpy as np

# Load the lostandfound dataset
lostandfound_semantic = load_dataset("BhavanaMalla/lostandfound-sub")

# Define the color map and class names
class_names = ["id", "ood", "background"]
palette = np.array([
    (5, 136, 245),  # ID - blues # (70, 130, 180)
    (220, 20, 60),   # OOD - red
    (0, 0, 0),       # Background - black
]) / 255.0

# Define ignored color
ignored_color = (0, 0, 0)

# Select one sample from the training dataset
sample = lostandfound_semantic["data"][15]  # 11, 15 Change index as needed
image = np.array(sample["image"])  # The input image
mask = np.array(sample["mask"])  # The ground truth mask

# Create the transparent blue overlay
alpha = 0.5  # Transparency level (0: fully transparent, 1: fully opaque)

# Initialize the overlay image
overlay = image.copy() / 255.0  # Normalize the RGB image

id_mask = (mask == 0)  # Where the GT mask has "ID" class
# Apply the transparent blue overlay
overlay[id_mask] = alpha * palette[0] + (1 - alpha) * overlay[id_mask]

ood_mask = (mask == 1)  # Where the GT mask has "OOD" class
overlay[ood_mask] = alpha * palette[1] + (1 - alpha) * overlay[ood_mask]

bg_mask = (mask == 255)  # Where the GT mask has "Background" class
overlay[bg_mask] = palette[2]

# Plot and save the image
plt.figure(figsize=(10, 8))
plt.imshow(image)
plt.axis("off")
plt.savefig("lostandfound_2.pdf", bbox_inches="tight", pad_inches=0)
plt.close()

# Plot and save the overlayed image
plt.figure(figsize=(10, 8))
plt.imshow(overlay)
plt.axis("off")
plt.savefig("lostandfound_gt_mask_2.pdf", bbox_inches="tight", pad_inches=0)
plt.close()

print("lostandfound image and gt_mask saved")


lostandfound image and gt_mask saved


In [ ]:
import torch
from collections import Counter
from datasets import load_dataset
import numpy as np
from torchvision.transforms.functional import pil_to_tensor

def compute_class_frequencies(hf_dataset, num_classes):
    """
    Compute per-class pixel frequency for a semantic segmentation test set.

    Args:
        num_classes (int): Total number of classes.

    Returns:
        class_frequencies (torch.Tensor): A tensor of shape (num_classes,) where each value
                                          represents the frequency of the corresponding class.
    """
    class_counts = torch.zeros(num_classes, dtype=torch.long)
    total_pixels = 0

    for sample in hf_dataset["train"]:
        mask = sample["annotation_labeltrainids"]  # Access the segmentation mask
        targets = pil_to_tensor(mask)
        # print(targets.shape) # Shape: (B, H, W),  torch.Size([1, 1024, 2048])

        # Deal background class
        background_mask = targets == 255
        targets[background_mask] = -1 # Set background pixels to -1 (or any ignored label value)
        # Filter out the background pixels
        targets = targets[targets != -1]  # Only keep non-background pixels
        targets = targets.view(-1) # Flatten to (N,)

        # Count pixel occurrences for each class
        class_counter = Counter(targets.tolist())
        for class_idx, count in class_counter.items():
            class_counts[class_idx] += count

        # Update total pixel count
        total_pixels += targets.numel()

    # Compute frequency as the fraction of pixels belonging to each class
    class_frequencies = class_counts.float() / total_pixels
    return class_frequencies


# Load the Cityscapes dataset
cityscapes_semantic = load_dataset("BhavanaMalla/cityscapes-semantic")

# Define the color map and class names
class_names = ["Road", "Sidewalk", "Building", "Wall", "Fence", "Pole", "Traffic light", "Traffic sign", "Vegetation", "Terrain", "Sky", "Person", "Rider", "Car", "Truck", "Bus", "Train", "Motorcycle", "Bicycle"]

num_classes = 19
class_frequencies = compute_class_frequencies(cityscapes_semantic, num_classes)

print("Train Per-Class Frequencies:", class_frequencies.tolist())


Train Per-Class Frequencies: [0.3688097894191742, 0.060868505388498306, 0.22819533944129944, 0.006559294182807207, 0.008782995864748955, 0.012276182882487774, 0.002084992593154311, 0.005528820212930441, 0.1591736078262329, 0.011586575768887997, 0.0401151068508625, 0.012173036113381386, 0.0013485692907124758, 0.07001111656427383, 0.00267634354531765, 0.002354058437049389, 0.002330171875655651, 0.0009864716557785869, 0.004139005206525326]


In [ ]:
import torch
from collections import Counter
from datasets import load_dataset
import numpy as np
from torchvision.transforms.functional import pil_to_tensor

def compute_class_frequencies(hf_dataset, num_classes):
    """
    Compute per-class pixel frequency for a semantic segmentation test set.

    Args:
        num_classes (int): Total number of classes.

    Returns:
        class_frequencies (torch.Tensor): A tensor of shape (num_classes,) where each value
                                          represents the frequency of the corresponding class.
    """
    class_counts = torch.zeros(num_classes, dtype=torch.long)
    total_pixels = 0

    for sample in hf_dataset["validation"]:
        mask = sample["annotation_labeltrainids"]  # Access the segmentation mask
        targets = pil_to_tensor(mask)
        # print(targets.shape) # Shape: (B, H, W),  torch.Size([1, 1024, 2048])

        # Deal background class
        background_mask = targets == 255
        targets[background_mask] = -1 # Set background pixels to -1 (or any ignored label value)
        # Filter out the background pixels
        targets = targets[targets != -1]  # Only keep non-background pixels
        targets = targets.view(-1) # Flatten to (N,)

        # Count pixel occurrences for each class
        class_counter = Counter(targets.tolist())
        for class_idx, count in class_counter.items():
            class_counts[class_idx] += count

        # Update total pixel count
        total_pixels += targets.numel()

    # Compute frequency as the fraction of pixels belonging to each class
    class_frequencies = class_counts.float() / total_pixels
    return class_frequencies


# Load the Cityscapes dataset
cityscapes_semantic = load_dataset("BhavanaMalla/cityscapes-semantic")

# Define the color map and class names
class_names = ["Road", "Sidewalk", "Building", "Wall", "Fence", "Pole", "Traffic light", "Traffic sign", "Vegetation", "Terrain", "Sky", "Person", "Rider", "Car", "Truck", "Bus", "Train", "Motorcycle", "Bicycle"]

num_classes = 19
class_frequencies = compute_class_frequencies(cityscapes_semantic, num_classes)

print("Per-Class Frequencies:", class_frequencies.tolist())


Per-Class Frequencies: [0.37661826610565186, 0.054065849632024765, 0.21916531026363373, 0.007331885863095522, 0.00821160338819027, 0.01479898951947689, 0.0019787719938904047, 0.006666385103017092, 0.17311407625675201, 0.0083194300532341, 0.03350081294775009, 0.01297159306704998, 0.0021497481502592564, 0.06519414484500885, 0.003011520719155669, 0.0038883714005351067, 0.0011259635211899877, 0.0007952151354402304, 0.007092074956744909]


"Road",
"Sidewalk",
"Building",
"Wall",
"Fence",
"Pole",
"Traffic light",
"Traffic sign",
"Vegetation",
"Terrain",
"Sky",
"Person",
"Rider",
"Car",
"Truck",
"Bus",
"Train",
"Motorcycle",
"Bicycle"

0.37661826610565186,
0.054065849632024765,
0.21916531026363373,
0.007331885863095522,
0.00821160338819027,
0.01479898951947689,
0.0019787719938904047,
0.006666385103017092,
0.17311407625675201,
0.0083194300532341,
0.03350081294775009,
0.01297159306704998,
0.0021497481502592564,
0.06519414484500885,
0.003011520719155669,
0.0038883714005351067,
0.0011259635211899877,
0.0007952151354402304,
0.007092074956744909

In [1]:
!pip install -U kaleido

In [6]:
import plotly.io as pio
pio.kaleido.scope.mathjax = None

In [7]:
import plotly.graph_objects as go
from plotly.io import write_image

classes = ["road", "building", "vegetation", "car", "sidewalk", "sky",
           "pole", "person", "terrain", "fence", "wall", "traffic_sign",
           "bicycle", "truck", "bus", "train", "traffic_light", "rider", "motorcycle"]
models = ["Baseline", "MCD", "OHEM", "EDL"]

# Per-class accuracies for each model
accuracies = {
    "Baseline": [val * 100 for val in [0.9919, 0.9644, 0.9677, 0.9763, 0.9185, 0.984,	0.7351,	0.8806,	0.7339,	0.6478,	0.6762,	0.8827,	0.8875,	0.8812,	0.9267,	0.8826,	0.8484,	0.7994,	0.8012]],
    "MCD": [val * 100 for val in [0.989, 0.9635, 0.9634, 0.9729, 0.9283, 0.9825, 0.7185, 0.9004, 0.7073, 0.692, 0.6166, 0.8692, 0.8797, 0.9072, 0.9285, 0.8841, 0.8366, 0.784 , 0.7988]],
    "OHEM": [val * 100 for val in [0.9924, 0.967, 0.9646, 0.9757, 0.9175, 0.9852, 0.7428, 0.9087, 0.7621, 0.6496, 0.6729, 0.8888, 0.8748, 0.8839, 0.926, 0.8855, 0.851, 0.7904, 0.7769]],
    "EDL": [val * 100 for val in [0.99022, 0.94938, 0.95181, 0.95944, 0.89432, 0.98095, 0.70584, 0.8545, 0.74203, 0.60448, 0.65576, 0.85708, 0.87761, 0.59648, 0, 0.81012, 0.00013563, 0.30716, 0.48497]],
}

# Create the figure
fig = go.Figure()

# Add a bar trace for each model
for model, acc in accuracies.items():
    fig.add_trace(go.Bar(
        x=classes,
        y=acc,
        name=model
    ))

# Customize layout
fig.update_layout(
    title="",
    xaxis_title="Classes",
    yaxis_title="Accuracy(%)",
    template="plotly_white",
    barmode='group',  # Grouped bar chart
    legend=dict(
        title="Models",
        bordercolor="gray",
        borderwidth=1,
        bgcolor="white",
        traceorder="normal", font=dict(size=14),
    ),
    xaxis=dict(
       showline=True, linewidth=1, linecolor='black',
        tickangle=-45, # Rotate class labels for readability
        title_font=dict(size=18),  # Increase xlabel font size
        tickfont=dict(size=14)  # Increase tick font size for x-axis
    ),
    yaxis=dict(
        showline=True, linewidth=1, linecolor='black',
        showgrid=True,
        gridcolor="lightgray",
        range=[0, 100],  # Y-axis range from 0 to 100
        dtick=10,  # Set custom tick interval to 20
        title_font=dict(size=18),  # Increase ylabel font size
        tickfont=dict(size=14)  # Increase tick font size for y-axis

    ),
    plot_bgcolor="white",  # Ensure a clean white background
    # paper_bgcolor="lightgray",  # Light gray border
    margin=dict(l=10, r=10, t=50, b=50)  # Margin inside the border
)

# Show the plot
fig.show()
write_image(fig=fig, file="per_class_accuracy_vanilla.pdf", format="pdf")


In [3]:
import plotly.graph_objects as go
from plotly.io import write_image

classes = ["road", "building", "vegetation", "car", "sidewalk", "sky",
           "pole", "person", "terrain", "fence", "wall", "traffic_sign",
           "bicycle", "truck", "bus", "train", "traffic_light", "rider", "motorcycle"]
models = ["Baseline", "MCD", "OHEM", "EDL"]

# Per-class accuracies for finetuned model
accuracies = {
    "Baseline": [val * 100 for val in [0.9919, 0.9644, 0.9677, 0.9763, 0.9185, 0.984,	0.7351,	0.8806,	0.7339,	0.6478,	0.6762,	0.8827,	0.8875,	0.8812,	0.9267,	0.8826,	0.8484,	0.7994,	0.8012]],
    "MCD": [val * 100 for val in [0.9926, 0.9688,	0.9656,	0.9779,	0.9177,	0.9858,	0.7372,	0.9011,	0.729,	0.6555,	0.6846,	0.8818,	0.8694,	0.893,	0.9261,	0.91, 0.8498, 0.8001, 0.7917]],
    "OHEM": [val * 100 for val in [0.9922,	0.9682,	0.9667,	0.9777,	0.9204,	0.9857,	0.7378,	0.9044,	0.7092,	0.6456,	0.6982,	0.8812,	0.8825,	0.8845,	0.9252,	0.8968,	0.8508,	0.7718,	0.7973]],
    "EDL": [val * 100 for val in [0.9903,	0.9592,	0.9592,	0.9737,	0.9147,	0.9822,	0.7239,	0.8918,	0.7317,	0.658,	0.7173,	0.8668,	0.8695,	0.882,	0.9262,	0.8729,	0.8421,	0.7347,	0.7647]],
}

# Create the figure
fig = go.Figure()

# Add a bar trace for each model
for model, acc in accuracies.items():
    fig.add_trace(go.Bar(
        x=classes,
        y=acc,
        name=model
    ))

# Customize layout
fig.update_layout(
    title="",
    xaxis_title="Classes",
    yaxis_title="Accuracy(%)",
    template="plotly_white",
    barmode='group',  # Grouped bar chart
    legend=dict(
        title="Finetuned <br> Models",
        bordercolor="gray",
        borderwidth=1,
        bgcolor="white",
        traceorder="normal", font=dict(size=14),
    ),
    xaxis=dict(
       showline=True, linewidth=1, linecolor='black',
        tickangle=-45, # Rotate class labels for readability
        title_font=dict(size=18),  # Increase xlabel font size
        tickfont=dict(size=14)  # Increase tick font size for x-axis
    ),
    yaxis=dict(
        showline=True, linewidth=1, linecolor='black',
        showgrid=True,
        gridcolor="lightgray",
        range=[0, 100],  # Y-axis range from 0 to 100
        dtick=10,  # Set custom tick interval to 20
        title_font=dict(size=18),  # Increase ylabel font size
        tickfont=dict(size=14)  # Increase tick font size for y-axis

    ),
    plot_bgcolor="white",  # Ensure a clean white background
    # paper_bgcolor="lightgray",  # Light gray border
    margin=dict(l=10, r=10, t=50, b=50)  # Margin inside the border
)

# Show the plot
fig.show()
write_image(fig=fig, file="per_class_accuracy_finetuned.pdf", format="pdf")


In [4]:
import plotly.graph_objects as go
from plotly.io import write_image

classes = ["road", "building", "vegetation", "car", "sidewalk", "sky",
           "pole", "person", "terrain", "fence", "wall", "traffic_sign",
           "bicycle", "truck", "bus", "train", "traffic_light", "rider", "motorcycle"]
models = ["Baseline", "MCD", "OHEM", "EDL"]

# Per-class iou for each model
ious = {
    "Baseline": [val * 100 for val in [0.9815,0.9257,0.9264,0.9487,0.8473,0.9492,0.6428,0.8143,0.6523,0.5815,0.5697,0.7906,0.7674,0.8201,0.8437,0.7821,0.6994,0.592,0.6096]],
    "MCD": [val * 100 for val in [0.9802,0.9244,0.9243,0.9474,0.8351,0.9485,0.6156,0.8161,0.6262,0.5582,0.5609,0.7822,0.7631,0.8202,0.8842,0.4408,0.6848,0.617,0.6277]],
    "OHEM": [val * 100 for val in [0.9816,0.9268,0.928,0.9508,0.8478,0.9502,0.6503,0.8279,0.6605,0.581,0.5884,0.7945,0.77,0.8301,0.8527,0.7442,0.7002,0.6206,0.6142]],
    "EDL": [val * 100 for val in [0.97908,0.91502,0.91765,0.93109,0.82501,0.93825,0.52804,0.7519,0.5594,0.49241,0.52323,0.68941,0.65874,0.31721,0,0.12541,0.00013299,0.26633,0.081485]],
}
# Create the figure
fig = go.Figure()

# Add a bar trace for each model
for model, iou in ious.items():
    fig.add_trace(go.Bar(
        x=classes,
        y=iou,
        name=model
    ))

# Customize layout
fig.update_layout(
    title="",
    xaxis_title="Classes",
    yaxis_title="IoU(%)",
    template="plotly_white",
    barmode='group',  # Grouped bar chart
    legend=dict(
        title="Models",
        bordercolor="gray",
        borderwidth=1,
        bgcolor="white",
        traceorder="normal", font=dict(size=14),
    ),
    xaxis=dict(
       showline=True, linewidth=1, linecolor='black',
        tickangle=-45, # Rotate class labels for readability
        title_font=dict(size=18),  # Increase xlabel font size
        tickfont=dict(size=14)  # Increase tick font size for x-axis
    ),
    yaxis=dict(
        showline=True, linewidth=1, linecolor='black',
        showgrid=True,
        gridcolor="lightgray",
        range=[0, 100],  # Y-axis range from 0 to 100
        dtick=10,  # Set custom tick interval to 20
        title_font=dict(size=18),  # Increase ylabel font size
        tickfont=dict(size=14)  # Increase tick font size for y-axis

    ),
    plot_bgcolor="white",  # Ensure a clean white background
    # paper_bgcolor="lightgray",  # Light gray border
    margin=dict(l=10, r=10, t=50, b=50)  # Margin inside the border
)

# Show the plot
fig.show()
write_image(fig=fig, file="per_class_iou_vanilla.pdf", format="pdf")


In [5]:
import plotly.graph_objects as go
from plotly.io import write_image

classes = ["road", "building", "vegetation", "car", "sidewalk", "sky",
           "pole", "person", "terrain", "fence", "wall", "traffic_sign",
           "bicycle", "truck", "bus", "train", "traffic_light", "rider", "motorcycle"]
models = ["Baseline", "MCD", "OHEM", "EDL"]

# Per-class iou for each model
ious = {
    "Baseline": [val * 100 for val in [0.9815,0.9257,0.9264,0.9487,0.8473,0.9492,0.6428,0.8143,0.6523,0.5815,0.5697,0.7906,0.7674,0.8201,0.8437,0.7821,0.6994,0.592,0.6096]],
    "MCD": [val * 100 for val in [0.9817,	0.9279,	0.9275,	0.9515,	0.8498,	0.9502,	0.6491,	0.8273,	0.6507,	0.5878,	0.6075,	0.7942,	0.7719,	0.8386,	0.8704,	0.8337,	0.7082,	0.6199,	0.6335]],
    "OHEM": [val * 100 for val in [0.9817,	0.9282,	0.9273,	0.9509,	0.8496,	0.9501,	0.6479,	0.8275,	0.6453,	0.5798,	0.6144,	0.7939,	0.7748,	0.8202,	0.8644,	0.8076,	0.7077,	0.6209,	0.6419]],
    "EDL": [val * 100 for val in [0.9807,	0.9236,	0.924,	0.9434,	0.8345,	0.944,	0.5969,	0.7965,	0.6383,	0.5559,	0.5846,	0.7537,	0.7302,	0.7784,	0.7779,	0.6526,	0.6286,	0.56,	0.3178]]
}

# Create the figure
fig = go.Figure()

# Add a bar trace for each model
for model, iou in ious.items():
    fig.add_trace(go.Bar(
        x=classes,
        y=iou,
        name=model
    ))

# Customize layout
fig.update_layout(
    title="",
    xaxis_title="Classes",
    yaxis_title="IoU(%)",
    template="plotly_white",
    barmode='group',  # Grouped bar chart
    legend=dict(
        title="Finetuned <br> Models",
        bordercolor="gray",
        borderwidth=1,
        bgcolor="white",
        traceorder="normal", font=dict(size=14),
    ),
    xaxis=dict(
       showline=True, linewidth=1, linecolor='black',
        tickangle=-45, # Rotate class labels for readability
        title_font=dict(size=18),  # Increase xlabel font size
        tickfont=dict(size=14)  # Increase tick font size for x-axis
    ),
    yaxis=dict(
        showline=True, linewidth=1, linecolor='black',
        showgrid=True,
        gridcolor="lightgray",
        range=[0, 100],  # Y-axis range from 0 to 100
        dtick=10,  # Set custom tick interval to 20
        title_font=dict(size=18),  # Increase ylabel font size
        tickfont=dict(size=14)  # Increase tick font size for y-axis

    ),
    plot_bgcolor="white",  # Ensure a clean white background
    # paper_bgcolor="lightgray",  # Light gray border
    margin=dict(l=10, r=10, t=50, b=50)  # Margin inside the border
)

# Show the plot
fig.show()
write_image(fig=fig, file="per_class_iou_finetuned.pdf", format="pdf")


In [4]:
!pip install -U kaleido

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.9/79.9 MB 28.4 MB/s eta 0:00:00


In [6]:
import plotly.io as pio
pio.kaleido.scope.mathjax = None

AttributeError: 'NoneType' object has no attribute 'mathjax'

In [1]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [2]:
# fishyscapes finetuned models roc data
!cp /content/drive/MyDrive/Colab\ Notebooks/Master\ Thesis/Results/roc_pkls/roc_result_0_Baseline.pkl .
!cp /content/drive/MyDrive/Colab\ Notebooks/Master\ Thesis/Results/roc_pkls/roc_result_0_EDL.pkl .
!cp /content/drive/MyDrive/Colab\ Notebooks/Master\ Thesis/Results/roc_pkls/roc_result_0_MCD.pkl .
!cp /content/drive/MyDrive/Colab\ Notebooks/Master\ Thesis/Results/roc_pkls/roc_result_0_MCD+OHEM.pkl .
!cp /content/drive/MyDrive/Colab\ Notebooks/Master\ Thesis/Results/roc_pkls/roc_result_1_EDL+UNC.pkl .
!cp /content/drive/MyDrive/Colab\ Notebooks/Master\ Thesis/Results/roc_pkls/roc_result_1_OHEM.pkl .
!cp /content/drive/MyDrive/Colab\ Notebooks/Master\ Thesis/Results/roc_pkls/roc_result_2_EDL+OHEM.pkl .

In [ ]:
import plotly.graph_objects as go
from plotly.io import write_image
import pickle

result_files = ["roc_result_0_Baseline.pkl", "roc_result_0_EDL.pkl", "roc_result_0_MCD.pkl",
#  "roc_result_1_OHEM.pkl",
                # "roc_result_0_MCD+OHEM.pkl", "roc_result_1_EDL+UNC.pkl", "roc_result_2_EDL+OHEM.pkl"
                ]

# Combine results into one figure
fig = go.Figure()
for result_file in result_files:
    # print(f"[INFO]: Opening the result file {result_file}")
    with open(result_file, 'rb') as f:
        print(f"[INFO]: Opening the result file {result_file}")
        roc_curve_data, auroc, fpr_at_recall_level, label = pickle.load(f)

        fpr, tpr = roc_curve_data[0], roc_curve_data[1]
        fig.add_trace(
            go.Scatter(
                x=fpr,
                y=tpr,
                mode='lines',
                name=f'{label}: AUROC={auroc:.3f}, FPR@95={fpr_at_recall_level:.3f}',
                showlegend=True,
            )
        )
        fig.add_trace(
            go.Scatter(
                x=[fpr_at_recall_level],
                y=[0.95],
                mode='markers',
                marker=dict(color='red', size=5),
                showlegend=False,
            )
        )

        # Update layout
        fig.update_layout(
            title="Combined ROC Curves for Multiple Datasets",
            xaxis=dict(
                title="False Positive Rate",
                showline=True, linewidth=1, linecolor="black", mirror=True,
                showgrid=True, gridcolor="lightgray",
            ),
            yaxis=dict(
                title="True Positive Rate",
                showline=True, linewidth=1, linecolor="black", mirror=True,
                showgrid=True, gridcolor="lightgray",
                title_font=dict(size=18),
                tickfont=dict(size=14),
                range=[0.8, 1]
            ),
            showlegend=True,
            legend=dict(
                title="<b>Legend</b>",
                x=0.99, y=0.01,
                xanchor="right", yanchor="bottom",
                bordercolor="gray",
                borderwidth=1,
                font=dict(size=12)
            ),
            plot_bgcolor="white",
            paper_bgcolor="white"
        )

fig.show()

[INFO]: Opening the result file roc_result_0_Baseline.pkl
3
trace added
Buffered data was truncated after reaching the output size limit.

In [12]:
range(len(a))

Buffered data was truncated after reaching the output size limit.